### ISSUE TITLE
Cache password verification in-memory

### ISSUE THREAD

**Issue by micw** [Mailu/Mailu#1411]
Currently our users must decide between strong password hashes and good system performance. The reason is that password hashes needs to be verified on every single request (e.g. on every click on webmail). With strong hashes, a single user can easily max-out multiple cpu cores.

Proposal:
- on the admin service, do a verification of passwords against the strong hash in the database.
- on successfull auth, cache the email + strong hash along with a weak/fast hash of the password in memory
- on each auth attempt, 1st check if there's a weak hash cached with the strong hash for that email address. If so, validate the password against the weak/fast hash instead.

With this change, password probe attempts are still slowed down. Passwords are still stored secure in database. Weak hashes are only cached in-memory. Auth for valid users will be speed up a lot.

Comments are welcome!

**Comment by kaiyou** [comment]
TL;DR: nice idea, lacking some touches, yet I fear the missing ingredients would fix the webmail whitelisting and render this irrelevant altogether.

I fail to see how this actually slows down password enumeration: it the password is currently cached for the user, an attacker will be able to run many attempts against the weak hash.

As a side (unimportant) note, it does not help against DOS if the attacker has the means to query from many IP addresses and thus bypass rate limiting (of course, such an attacker would have plenty other DOS options, so I guess it is irrelevant).

In the end, here is my opinion : because we use rate-limiting and are mostly protecting ourselves against online password attacks, I do not believe that cpu-heavy hashes are the most important requirement, except for the current weak point that is the webmail (rate-limite whitelisted).

If we really wish to add offline attacks to the equation (from a threat modelling perspective, I would argue that having your email compromised is way worse that having your plaintext password recovered from the has, but still, let's offer more security if we can do it at low cost), your approach is probably on the right track although lacking some touches I am afraid. We should look around for other people that solved identical problems.

One would want to add the source IP as a cache key, but then they would bump into the webmail issue. So we could solve the webmail issue altogether, but then I think rate limiting would suffice and we could drop the requirement on cpu-heavy hashes.

**Comment by kaiyou** [comment]
As a (kind of relevant) side note: this could provide an attacker with an oracle about online/offline users.

**Comment by micw** [comment]
Thank you for feedback. So I'd change to: verify against weak hash, on failure verify against strong (or delay).

**Comment by kaiyou** [comment]
It still does not protect: the attacker has no reason to wait for the delay or the strong hash to compute, they only needs to try another password if they do not get a positive feedback quickly.

**Comment by Nebukadneza** [comment]
Hi There,

The `Mailu`-Project is currently in a bit of a bind! We are short on man-power, and we need to judge if it is possible for us to put in some work on this issue.

To help with that, we are currently trying to find out which issues are actively keeping users from using `Mailu`, which issues have someone who want to work on them — and which issues may be less important. These a less important ones could be discarded for the time being, until the project is in a more stable and regular state once again.

In order for us to better assess this, it would be helpful if you could put a **reaction on this post** (use the :smiley: icon to the top-right).
- 👍️ if you **need this** to be able to use Mailu. Ideally, you’d also be able to test this on your installation, and provide feedback …
- 🎉 if you find it a **nice bonus**, but no deal-breaker
- 🚀 if you want to **work on it yourself**!
We want to keep this **voting open for 2 weeks** from now, so please help out!

**Comment by nextgens** [comment]
> One would want to add the source IP as a cache key, but then they would bump into the webmail issue. So we could solve the webmail issue altogether, but then I think rate limiting would suffice and we could drop the requirement on cpu-heavy hashes.

I've implemented something in #1745 to that effect (a per-ip temporary rate limit exemption).

IMHO the right solution when the protocol is HTTP (webmail, admin interface, ...) is to challenge the user with a captcha... and forget about IP-based rate limits.

**Comment by micw** [comment]
Ideally, webmail should use internal endpoints, bypass rate-limiting and should (as you suggest) to the rate-limiting on http based on log-in failures.

**Comment by nextgens** [comment]
> Ideally, webmail should use internal endpoints

I'm not sure I am following here... sure they could for authentication... but then they would still need IMAP/dovecot access for the mailboxes, no?

What would be the pros of doing it that way?

**Comment by nextgens** [comment]
@micw I've now ensured that both webmails use the internal endpoint in #1745 ... it seems to work. Please test/review

**Comment by nextgens** [comment]
Passlib has deprecated/removed min_verify_time https://foss.heptapod.net/python-libs/passlib/-/issues/83 their rationale for doing so applies here too

**Comment by nextgens** [comment]
> Ideally, webmail should use internal endpoints, bypass rate-limiting and should (as you suggest) to the rate-limiting on http based on log-in failures.

This is implemented in #1754
